### PR TITLE
Draft review guidelines proposal

### PR DESCRIPTION
**Pull request by DRMacIver**
This is an attempt at solving #505 - it adds a section to the CONTRIBUTING.rst outlining what we should be looking for in review.

Critiques and suggested additions *extremely* welcome, both from Hypothesis maintainers and externally.

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -162,3 +162,134 @@ their individual contributions.
...
+
+Public API changes require the most careful scrutiny of all reviews,
+because they are the ones we are stuck with for the longest: Hypothesis
+follows semver and additionally tries not to break API compatibility
+wherever possible.
+
+When a public API changes we should ask the following questions:
+
+1. Is this change backwards compatible?
+2. Are ther eany backwards incompatible changes we'd want to make to this
```
Comment (alexwlchan) [inline CONTRIBUTING.rst]: /s/ther eany/there any

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -162,3 +162,134 @@ their individual contributions.
...
+* The bug in the understanding of the author when the bug was introduced
+* The bug in the testing that lead to this bug not being caught
+
+Not every bug fix has to address all three bugs, but code review should
+at least keep an eye on the second two and see if there is anything
+more general we could or should be doing to fix the others.
+
+In particular a reviewer should ask the following:
+
+1. Does the testing for this adequately capture the bug? t's OK for some
```
Comment (alexwlchan) [inline CONTRIBUTING.rst]: /s/t's/It's

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -162,3 +162,134 @@ their individual contributions.
...
+follows semver and additionally tries not to break API compatibility
+wherever possible.
+
+When a public API changes we should ask the following questions:
+
+1. Is this change backwards compatible?
+2. Are ther eany backwards incompatible changes we'd want to make to this
+   API later and if so can we do them now instead? 
+3. How does this API handle invalid input and are the error messages clear
+   when that happens?
```
Comment (alexwlchan) [inline CONTRIBUTING.rst]: You and I both have a policy that error messages should include the bad input, if/when it occurs. Worth stating here as an explicit principle?

File: CONTRIBUTING.rst (hunk 2 above)
Comment (adamtheturtle) [inline CONTRIBUTING.rst]: t's > It's

File: CONTRIBUTING.rst (hunk 3 above)
Comment (DRMacIver) [inline CONTRIBUTING.rst]: Yes, definitely.

**Comment by DRMacIver** [comment]
> I’m worried it might be a little “wall of text” and lacking in concrete guidance, just lots of questions, but I need to think about that more.

Yeah I'm a bit worried about this too but I'm not sure what to do about it.

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -162,3 +162,136 @@ their individual contributions.
...
+   when that happens clear?
+5. Are there any cases where this API could fail silently and how can
+   those be avoided?
+6. Could this API introduce a source of flakiness? Flakiness here is
+   defined as anything where Hypothesis fails and this does not indicate
+   a bug in Hypothesis or in the way the user wrote the code or the test.
+7. Will this API be hard to support in the long-term? In particular is the
+   ability to support it very tied to the current Hypothesis internals?
+8. Is any new API adequately documented?
+9. Is the upgrade path clear for any deprecated API?
```
Comment (adamtheturtle) [inline CONTRIBUTING.rst]: What, roughly, constitutes a clear upgrade path?

Comment (adamtheturtle) [inline CONTRIBUTING.rst]: Maybe see https://pyopenssl.readthedocs.io/en/latest/backward-compatibility.html as an example of guidelines for backwards incompatible changes.

Comment (DRMacIver) [inline CONTRIBUTING.rst]: I've updated the wording. Does that make it clearer? Reading the link now.

**Comment by Zac-HD** [comment]
I think we need to decide whether this is a *descriptive* overview of what makes a good pull request, or a *prescriptive* checklist of requirements.  If we want a checklist - personally I prefer minimal process but the literature is pretty convincing - it should be a concise list of questions with unambiguous yes / no / not applicable answers and a wrong answer should be grounds to block merging.  In fact, the bit about a concise list of questions with objective answers applies regardless!

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -162,3 +162,138 @@ their individual contributions.
...
+~~~~~~~~~~~~~~~~~
+General Questions
+~~~~~~~~~~~~~~~~~
+
+The following are useful questions to ask for almost every change:
+
+1. Is this change too large? Could it profitably be broken up into
+   several smaller changes that could be judged on their individual
+   merits?
+2. Is the motivation for this change clear?
```
Comment (Zac-HD) [inline CONTRIBUTING.rst]: Clear to who?  Suggest "Has the motivation for this change been clearly explained?", with a target audience of non-DRMacIver maintainers (generally familiar with Hypothesis, but not assuming deep knowledge of internals).

Then again, I'm basically saying "explain this to me please" so it's not an unbiased requirement 😉

**Comment by DRMacIver** [comment]
I have pushed an update that makes it drastically more prescriptive and makes it more checklisty. @Zac-HD, @alexwlchan, what do you think?

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -162,3 +162,150 @@ their individual contributions.
...
+The rest of this section outlines specific things reviewers should
+focus on in aid of this, broken up by sections according to their area
+of applicability.
+
+All of these conditions must be satisfied for merge. Where the reviewer
+thinks this conflicts with the above higher level goals, they may make
+an exception if both the author and another maintainer agree.
+
+~~~~~~~~~~~~~~~~~
+General Questions
```
Comment (bukzor) [inline CONTRIBUTING.rst]: Given the new prescriptive language, maybe "Requirements" is a better heading.

File: CONTRIBUTING.rst [hunk 7]
```diff
@@ -162,3 +162,150 @@ their individual contributions.
...
+
+This section applies to any changes in Hypothesis's behaviour, regardless
+of their nature.
+
+1. The code should be clear in its intent and behaviour.
+2. Behaviour changes should come with appropriate tests to demonstrate
+   the new behaviour.
+3. Hypothesis must never be *flaky*. Flakiness here is
+   defined as anything where a test fails and this does not indicate
+   a bug in Hypothesis or in the way the user wrote the code or the test.
```
Comment (bukzor) [inline CONTRIBUTING.rst]: How would this apply to a (hypothetical) pull request adding the assume() function?
There's some probability that assume() fails to find enough examples via random chance.

**Comment by Zac-HD** [comment]
The new version looks great to me.

File: CONTRIBUTING.rst (hunk 7 above)
Comment (DRMacIver) [inline CONTRIBUTING.rst]: That's true. I think the answer might be that assume shouldn't have been added without fixing that problem, and has been grandfathered in for now but should be made nonflaky and then that shouldn't be allowed to revert.

File: CONTRIBUTING.rst [hunk 8]
```diff
@@ -162,3 +162,150 @@ their individual contributions.
...
+   it doesn't count as public API!
+2. Changes must be backwards compatible. Where this is not possible, they
+   must first introduce a deprecation warning, then once the major version
+   is bumped the deprecation warning and the functionality may be removed.
+3. If an API is deprecated, the deprecation warning must make it clear
+   how the user should modify their code to adapt to this change (
+   possibly by referring to documentation).
+4. If it is likely that we will want to make backwards compatible changes
+   to an API later, to whatever extent possible these should be made immediately
+   when it is introduced instead.
```
Comment (alexwlchan) [inline CONTRIBUTING.rst]: Surely it’s more important to do this for backwards incompatible changes? If a future change is backwards compatible, doing it later doesn’t matter.

Comment (DRMacIver) [inline CONTRIBUTING.rst]: yes, that's just a typo. Whoops.

**Comment by DRMacIver** [comment]
> I’ve also only just noticed I never added myself to the list of contributors. Oops

I was thinking we might want to factor out a list of maintainers above and separate from the contributors list.

**Comment by alexwlchan** [comment]
> I was thinking we might want to factor out a list of maintainers above and separate from the contributors list.

Sounds like a good idea.

**Comment by DRMacIver** [comment]
After discussion on #515 I have added a section about settings.

Also: I think this is getting too long to be in a CONTRIBUTING.rst and would like to break it out into its own file. What do people think?

**Comment by Zac-HD** [comment]
A separate file makes sense to me.  I'd like to keep (duplicate?) the introductory text through to "The rest of this section outlines specific things..." and then link to details in another file.

File: CONTRIBUTING.rst [hunk 9]
```diff
@@ -85,8 +86,8 @@ You can also use 'make check-format', which will run format and some linting and
 git diff. Note: This will error even if you started with a git diff, so if you've got any uncommitted changes
 this will necessarily report an error.
 
-'make check' will run check-format and all of the tests. Warning: This will take a *very* long time. On travis the
-currently takes multiple hours of total build time (it runs in parallel on Travis so you don't have to wait
+'make check' will run check-format and all of the tests. Warning: This will take a *very* long time. On Travis the
+currently takes more than an hour of total build time (it runs in parallel on Travis so you don't have to wait
```
Comment (alexwlchan) [inline CONTRIBUTING.rst]: "On Travis the currently" doesn't make sense.

File: guides/review.rst [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+1. The code should be clear in its intent and behaviour.
+2. Behaviour changes should come with appropriate tests to demonstrate
+   the new behaviour.
+3. Hypothesis must never be *flaky*. Flakiness here is
+   defined as anything where a test fails and this does not indicate
+   a bug in Hypothesis or in the way the user wrote the code or the test.
+4. The version number must be kept up to date. If you're making a change,
+   it should increase the appropriate version (patch for things that don't
+   change public facing functionality, minor for things that do).
```
Comment (alexwlchan) [inline guides/review.rst]: Worth mentioning semver?

**Comment by DRMacIver** [comment]
>  would be a good place to put stuff about Hypothesis internals as well.

Yup. Obviously out of scope for this pull request, but that's very much the idea.

**Comment by DRMacIver** [comment]
Things from the recent build wrangling I'd like reference to in here but don't currently know quite what to say:

* Guidelines for build changes
* A note that I will sometimes use Admin fiat to bypass the system but will generally only do it when there is a really good reason (e.g. this build wrangling stuff is fairly pointless to try to push through a formal review and build) and people should call me out on it if they think I'm wrong to do so.

**Comment by DRMacIver** [comment]
What do we think of the current state of this? My inclination is good enough to merge though will need refinement over time. @alexwlchan, did I address all the changes you wanted?

**Comment by Zac-HD** [comment]
IMO "now is better than never" and we'll iterate regardless.  Merge away, after you fix the merge conflict in lint commands.  (I actually keep thinking this has already been merged...)

**Comment by DRMacIver** [comment]
Right. I've rebased this against master. Will wait for the build to run and then merge.

**Pull request merged**